src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp

```diff
@@ -78,7 +84,11 @@ struct ReadTakeCommand
         assert(0 <= remaining_samples_);
 
         current_slot_ = data_values_.length();
-        finished_ = nullptr == instance.second;
+        finished_ = nullptr == instance.second || remaining_samples_ == 0;
+        if (remaining_samples_ == 0)
+        {
+            return_value_ = ReturnCode_t::RETCODE_OK;
```

Comment:
Is there a test covering this case?

Reply:
[this one](https://github.com/eProsima/Fast-DDS/blob/3df6020d782f9d585a75e13c7ec13938809a97dd/test/unittest/dds/subscriber/DataReaderTests.cpp#L1068)

In fact, I added this bit of code because this test was failing on `i == 0`

Reply:
The idea is that, when called with `max_samples == 0`, read or take will return `NO_DATA` or `OK` depending on whether there are samples on the reader history for the expected status filters. If we are adding and removing samples on the collection (due to validity checks), I think it is fine to return OK, as there is no postcondition on the standard or our Doxygen documentation stating that `OK` implies `length > 0`.